integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/NotPredicateSpecificsIT.java

```diff
@@ -0,0 +1,152 @@
+import org.junit.Test;
+
+public class NotPredicateSpecificsIT {
+
+	private static final String DOCUMENT_1 = "1";
+	private static final String DOCUMENT_2 = "2";
+	private static final String DOCUMENT_3 = "3";
+
+	// Document 1
+
+	private static final String FIELD1_VALUE1 = "Irving";
+	private static final Integer FIELD2_VALUE1 = 3;
+	private static final Integer FIELD3_VALUE1 = 4;
+	private static final Integer FIELD4_VALUE1AND2 = 1_000;
+	private static final Integer FIELD5_VALUE1AND2 = 2_000;
+
+	// Document 2
+
+	private static final String FIELD1_VALUE2 = "Auster";
+	private static final Integer FIELD2_VALUE2 = 13;
+	private static final Integer FIELD3_VALUE2 = 14;
+	// Field 4: Same as document 1
+	// Field 5: Same as document 1
+
+	// Document 3
+
+	private static final String FIELD1_VALUE3 = "Coe";
+	private static final Integer FIELD2_VALUE3 = 25;
+	private static final Integer FIELD3_VALUE3 = 42;
+	private static final Integer FIELD4_VALUE3 = 42_000; // Different from document 1
+	private static final Integer FIELD5_VALUE3 = 142_000; // Different from document 1
+
+	@ClassRule
+	public static final SearchSetupHelper setupHelper = new SearchSetupHelper();
+
+	private static final SimpleMappedIndex<IndexBinding> index = SimpleMappedIndex.of( IndexBinding::new );
+
+	@BeforeClass
+	public static void setup() {
+		setupHelper.start().withIndex( index ).setup();
+
+		initData();
+	}
+
+	@Test
+	public void matchNone() {
+		assertThatQuery( index.query()
+				.where( f -> f.not( f.matchAll() ) ) )
+				.hasNoHits();
+	}
+
+	@Test
+	public void matchAll() {
+		assertThatQuery( index.query()
+				.where( f -> f.not( f.matchNone() ) ) )
+				.hasTotalHitCount( 3 )
+				.hasDocRefHitsAnyOrder( index.typeName(), DOCUMENT_1, DOCUMENT_2, DOCUMENT_3 );
+	}
+
+	@Test
+	public void notMatchSpecificValue() {
+		assertThatQuery( index.query()
+				.where( f -> f.not( f.match().field( "field2" ).matching( FIELD2_VALUE1 ) ) ) )
+				.hasTotalHitCount( 2 )
+				.hasDocRefHitsAnyOrder( index.typeName(), DOCUMENT_2, DOCUMENT_3 );
+	}
+
+	@Test
+	public void mustWithNotInside() {
+		assertThatQuery( index.query()
+				.where( f -> f.bool().must( f.not( f.match().field( "field2" ).matching( FIELD2_VALUE1 ) ) ) ) )
+				.hasTotalHitCount( 2 )
+				.hasDocRefHitsAnyOrder( index.typeName(), DOCUMENT_2, DOCUMENT_3 );
+	}
+
+	@Test
+	public void tooMuchNestedNot() {
+		SearchPredicateFactory f = index.createScope().predicate();
+		BooleanPredicateClausesStep<?> step = f.bool().must( f.not( f.not( f.not( f.not( f.not( f.not( f.not( f.match().field( "field2" ).matching( FIELD2_VALUE1 ) ) ) ) ) ) ) ) );
+		assertThatQuery( index.query()
+				.where( step.toPredicate() ) )
+				.hasTotalHitCount( 2 )
+				.hasDocRefHitsAnyOrder( index.typeName(), DOCUMENT_2, DOCUMENT_3 );
+
+		assertThatQuery( index.query()
+				.where( f.not( ( step ) ).toPredicate() ) )
+				.hasTotalHitCount( 1 )
+				.hasDocRefHitsAnyOrder( index.typeName(), DOCUMENT_1 );
```

Comment:
Could you also add some tests that check optimizations? Of course these would be backend-specific. You can check the content of a `SearchQuery` with `SearchQuery.queryString()`.

Reply:
Got it, will do 😃